### ISSUE TITLE
NVMe driver: admin queue does not update shadow doorbells

### ISSUE THREAD

**Issue by antroseco** [spdk/spdk#2147]
# Bug report

## Expected Behavior

When the host submits a command to the admin queue, it should update the admin queues shadow doorbell (if configured).

In the NVMe 2.0a Base Specification, Figure 167 clearly shows the admin queue doorbells whilst §B.5.1 states that
> After the completion of the Doorbell Buffer Config command, host software shall submit commands by updating the appropriate entry in the Shadow Doorbell buffer instead of updating the controller's corresponding doorbell property.

## Current Behavior

SPDK will correctly update the shadow doorbells of all IO queues, but not the shadow doorbells of the admin queue.

This is consistent with the Linux NVMe driver, but contradicts the spec. Is there a reason for this behaviour?

**Comment by changpe1** [comment]
There is no specific reason for that, just because ADMIN commands will not be sent from VM for most of the time.

**Comment by jlevon** [comment]
@benlwalker we're curious about the spec on this issue. Neither Linux nor spdk seems to follow the spec here, and it seems like we're forced to not follow it too

**Comment by benlwalker** [comment]
While I am reasonably confident that not updating the shadow doorbell for admin commands will work just fine, my reading of the specification is that we should be using the shadow doorbells for all queues if they are configured.

But let's look through the code here and figure out why it isn't occurring for the admin queue. I see admin queue commands get submitted through the same code path as I/O commands and ultimately still call `nvme_pcie_qpair_ring_sq_doorbell`. That function is checking if there is a shadow doorbell and using it if there is one. I assume this must be false for the admin queue.

The `has_shadow_doorbell` flag is currently set in the completion handler when a submission queue is created. But for PCIe devices, an admin queue isn't created via commands. It's created via MMIO at start up, so it simply never goes down the code path that would have set the flag. I think that's the only reason it isn't using the shadow doorbell.

The hard part is that shadow doorbells are PCIe specific, but in the admin queue creation path there isn't a call into the PCIe transport to go configure them currently. But I took a shot at it in the patch below. I don't have any way to test this.

https://review.spdk.io/gerrit/c/spdk/spdk/+/9502

**Comment by jimharris** [comment]
Hey Ben, I left a comment in the 9502 patch.  The general approach (hooking the completion of the DOORBELL CONFIG command) is correct, but the current version of the patch only does it in the multi-process case where the completion is deferred because a different process found the completion entry when it was polling the adminq.

Also a note for @changpe1 - I noticed during code inspection that we only try to configure the doorbells when the trtype == PCIE - is there any reason we couldn't use shadow doorbells when using the SPDK driver for VFIO_USER as well?

**Comment by jlevon** [comment]
So, the reason we were asking about this, is because there's a problem on the controller side. Where should a controller look for the admin queue's doorbell? We'd like to look in the shadow doorbell region, but we can't - it'll always be zero. We have to basically hard-code that for admin queues, we always have to check bar0 location *instead*.

In other words, we don't see a way in which we can implement an nvme controller that supports shadow doorbells without *assuming* the driver is not compliant.

In that sense, implementing the change above would completely break us in fact :)

**Comment by benlwalker** [comment]
Is it possible to map the doorbells so that you get an interrupt if they're written to, only when shadow doorbells are enabled? That way writing to the shadow doorbell is fast but you still get notified when the user writes to the actual doorbell.

But with vfio-user, is there any reason at all to be using shadow doorbells? Can't the guest just map the actual doorbells into shared memory? If you're using this with vfio-user, I'd just not support shadow doorbells.

**Comment by jimharris** [comment]
> Is it possible to map the doorbells so that you get an interrupt if they're written to, only when shadow doorbells are enabled? That way writing to the shadow doorbell is fast but you still get notified when the user writes to the actual doorbell.
> 
> But with vfio-user, is there any reason at all to be using shadow doorbells? Can't the guest just map the actual doorbells into shared memory? If you're using this with vfio-user, I'd just not support shadow doorbells.

You can map the actual doorbells into shared memory, but doesn't it still result in a trap because the guest is writing to the PCI BAR?  With shadow doorbells it avoids any MMIO writes to the virtual PCI device.

**Comment by jlevon** [comment]
Right, the setup we need/want is that the shadow doorbell region is just guest memory (no MMIO writes, no notifications possible), and BAR0 is mapped into the guest only via mmio (either REGION_WRITE or ioeventfd/ioregionfd).

> If you're using this with vfio-user, I'd just not support shadow doorbells

If we want interrupt mode to work, we want shadow doorbells so we don't pay the price of MMIO writes on every doorbell write.

**Comment by benlwalker** [comment]
So even though the "BAR" here is just some system memory we allocated in the target, because vfio-user presented it as a BAR then every access to it traps as if it were MMIO? I know there's a way to control whether there are notifications or not when the memory is written to, but changing that doesn't also change the trap behavior?

**Comment by changpe1** [comment]
> Hey Ben, I left a comment in the 9502 patch. The general approach (hooking the completion of the DOORBELL CONFIG command) is correct, but the current version of the patch only does it in the multi-process case where the completion is deferred because a different process found the completion entry when it was polling the adminq.
> 
> Also a note for @changpe1 - I noticed during code inspection that we only try to configure the doorbells when the trtype == PCIE - is there any reason we couldn't use shadow doorbells when using the SPDK driver for VFIO_USER as well?

Currently SPDK vfio-user is using sparse memory region for BAR0, that means the doorbells are already in the memory mapped address, there is no need to support shadow doorbell buffer config command.

**Comment by changpe1** [comment]
@jimharris @benlwalker For BAR0, we are using the sparse mmap way, that means for the first 4KiB, QEMU will trap and then translate the trap with socket messages, while for the next doorbells, it uses the memory mmap way, that means even without shadow doorbell support, it will not have traps for this memory region.

**Comment by jimharris** [comment]
> > Hey Ben, I left a comment in the 9502 patch. The general approach (hooking the completion of the DOORBELL CONFIG command) is correct, but the current version of the patch only does it in the multi-process case where the completion is deferred because a different process found the completion entry when it was polling the adminq.
> > Also a note for @changpe1 - I noticed during code inspection that we only try to configure the doorbells when the trtype == PCIE - is there any reason we couldn't use shadow doorbells when using the SPDK driver for VFIO_USER as well?
> 
> Currently SPDK vfio-user is using sparse memory region for BAR0, that means the doorbells are already in the memory mapped address, there is no need to support shadow doorbell buffer config command.

Got it - thanks @changpe1!

**Comment by jlevon** [comment]
just to be clear, we definitely want shadow doorbells for interrupt mode, or we are paying the MMIO write cost for every doorbell write.
